File: packages/codex-rotate-app/paths.ts
import { homedir } from "node:os";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const MODULE_DIR = dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = resolve(MODULE_DIR, "..", "..");

export interface CodexRotateAppPaths {
  repoRoot: string;
  codexHome: string;
  codexAuthFile: string;
  codexLogsDbFile: string;
  codexRotateEntrypoint: string;
  rotateAppHome: string;
  debugProfileDir: string;
  sessionFile: string;
  runtime: string;
}

export function resolveCodexRotateAppPaths(): CodexRotateAppPaths {
  const codexHome = process.env.CODEX_HOME ?? join(homedir(), ".codex");
  const rotateAppHome = join(homedir(), ".codex-rotate-app");
  return {
    repoRoot: REPO_ROOT,
    codexHome,
    codexAuthFile: join(codexHome, "auth.json"),
    codexLogsDbFile: join(codexHome, "logs_1.sqlite"),
    codexRotateEntrypoint: join(REPO_ROOT, "packages", "codex-rotate", "index.ts"),
    rotateAppHome,
    debugProfileDir: join(rotateAppHome, "profile"),
    sessionFile: join(rotateAppHome, "session.json"),
    runtime: process.env.CODEX_ROTATE_APP_RUNTIME
      ?? (process.versions.bun ? process.execPath : "bun"),
  };
}

File: packages/codex-rotate-app/crates/codex-rotate-tray-core/src/paths.rs
use std::path::PathBuf;

use anyhow::{Context, Result};

#[derive(Clone, Debug)]
pub struct TrayPaths {
    pub codex_auth_file: PathBuf,
    pub codex_logs_db_file: PathBuf,
    pub rotate_app_home: PathBuf,
    pub debug_profile_dir: PathBuf,
    pub session_file: PathBuf,
}

pub fn resolve_paths() -> Result<TrayPaths> {
    let home = dirs::home_dir().context("Failed to resolve home directory.")?;
    let codex_home = std::env::var_os("CODEX_HOME")
        .map(PathBuf::from)
        .unwrap_or_else(|| home.join(".codex"));
    let rotate_app_home = home.join(".codex-rotate-app");
    Ok(TrayPaths {
        codex_auth_file: codex_home.join("auth.json"),
        codex_logs_db_file: codex_home.join("logs_1.sqlite"),
        debug_profile_dir: rotate_app_home.join("profile"),
        session_file: rotate_app_home.join("session.json"),
        rotate_app_home,
    })
}
